### PR TITLE
Rework design by introspection

### PR DESCRIPTION
**Pull request by teodutu**
This PR reworks the content of [session 6 - Design by Introspection](https://ocw.cs.pub.ro/courses/dss/laboratoare/06).

Tasks:
- [x] Skeleton and solution
- [x] Argue in favour of logging
- [x] Logs for basic types
- [x] `static if`
- [x] UFCS
- [x] Logs for arrays
- [x] Logs for structs and classes
- [x] `__traits`
- [x] CTFE
- [x] String mixins
- [x] Duck Typing (optional)
- [x] UDAs (optional)

### CODE REVIEW

File: lab-06/README.md [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+
+**Hint:** Use `std.conv.to` to convert numeric types to strings.
+
+Write unittests for the new `log` functions.
+
+### Recap: UFCS
+
+Rememeber [UFCS](../lab-01/README.md#functions) from session 1.
+UFCS stands for Uniform Function Call Syntax and allows us to call function `foo` either as `foo(a)` or as `a.foo()`.
+This feature makes coode far more expressive.
```
Comment (RazvanN7) [inline lab-06/README.md:70]:
```suggestion
This feature makes code far more expressive.
```

File: lab-06/quiz/template-specialisations.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
+## Question Text
+
+Why cannot we use template specialisations to aggregate all numeric types?
```
Comment (RazvanN7) [inline lab-06/quiz/template-specialisations.md:3]: I think this question is a bit odd. You could technically use template specializations, but then you would have to have an overload for each tree of types. That makes both the second and the third answers vaguely correct. Maybe you should spell out the fact that you want a single template declarations and not multiple overloads.

File: lab-06/README.md [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+Iterate this list using a `foreach` statement and append the name of each member of the structure to the output string.
+
+[Quiz](./quizzes/compile-time-reflection.md)
+
+### Recap: CTFE
+
+We can otpimise the `foreach` loop we've just written at compile time and have the compiler handle the iteration.
+We can do this using [Compile Time Function Evaluation](https://tour.dlang.org/tour/en/gems/compile-time-function-evaluation-ctfe).
+We talked about this in [session 2](../lab-02/README.md#ctfe).
+Go over this material for a recap.
```
Comment (RazvanN7) [inline lab-06/README.md:232]: This sentence can be dropped.

File: lab-06/README.md [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+But there's a B-list, too, that includes paradigms such as logic, constraint-oriented, and symbolic.
+The point is, there aren't very many of them altogether.
+
+**Design by Introspection (DBI)** is a proposed programming paradigm that is at the same time explosively productive and firmly removed from any of the paradigms considered canon.
+The tenets of Design by Introspection are:
+1. **The rule of optionality:** Component primitives are almost entirely opt-in.
+A given component is required to implement only a modicum of primitives, and all others are optional.
+The component is free to implement any subset of the optional primitives.
+1. **The rule of introspection:** A component user employs introspection on the component to implement its own functionality using the primitives offered by the component.
+1. **The rule of elastic composition:** A component obtained by composing several other components offers capabilities in proportion with the capabilities offered by its individual components.
```
Comment (RazvanN7) [inline lab-06/README.md:15]: These 3 things are very vague and kind of tern. Maybe you should add an example for each. Also, it seems that we talk theoretically about design by introspection, but we don't really back it up with examples and useful patterns. I see that the logger application is interesting, but still, it's not really emphasized on design by introspection. Maybe we can rename this to course to "Metaprogramming"?

**Review by RazvanN7 — CHANGES_REQUESTED** [review]

**Comment by RazvanN7** [comment]
Also, what happened to `static if`? Did I miss something?

**Comment by RazvanN7** [comment]
I think that the narrative of this lab is a bit backwards. We should start with a recap of compile time evaluation and make sure that students understand that some things happen at compile time, whereas other things happen at run time. Then, as we progress we can explain templates, constraints, traits, static if. I am bit in doubt about the logger thing. I mean it's definitely useful, but I feel it stops us from having more compelling and simple examples.

**Comment by teodutu** [comment]
> Also, what happened to `static if`? Did I miss something?

It's covered in [lab 2](https://ocw.cs.pub.ro/courses/dss/laboratoare/02#static_if). I'll add a recap section about it since lab 2 is more than 1 week away from lab 6.

**Comment by teodutu** [comment]
> We should start with a recap of compile time evaluation and make sure that students understand that some things happen at compile time, whereas other things happen at run time. Then, as we progress we can explain templates, constraints, traits, static if.

Sounds good. I chose this narrative so each concept is introduced when it's needed, but the bigger picture that CTFE does exist + demos about it are definitely welcome at the beginning.

**Comment by teodutu** [comment]
> I think that the theoretical part can be fairly separated from the logger thing. What I mean is: present the concepts in a vacuum and then after each one tie it to the logger example.

Doesn't this run the risk of decoupling the theoretical concepts from the context in which they're needed? The current narrative wants to start from a use case for each of them so that students feel they're more useful.

**Comment by RazvanN7** [comment]
> Doesn't this run the risk of decoupling the theoretical concepts from the context in which they're needed? The current narrative wants to start from a use case for each of them so that students feel they're more useful.

Not as I see it. My perspective is that you introduce each concept and talk about it with some examples and then you exemplify it on the logger example. Baby steps.

**Comment by teodutu** [comment]
> My perspective is that you introduce each concept and talk about it with some examples and then you exemplify it on the logger example. Baby steps.

It makes sense. I'll do this.

File: lab-06/README.md (hunk 4 above)
Comment (teodutu) [reply lab-06/README.md:15]: I agree we should rename it to "Metaprogramming". Using some more advanced patterns would be cool, but I think that would require that this session be held in advance. Food for thought for next year, maybe?

**Review by RazvanN7 — APPROVED** [review]
Looks good! One last thing to cater for: @edi33416 has managed to create a website where our content is rendered (quizes work too!). Please make sure you follow the pattern in this PR: https://github.com/Dlang-UPB/D-Summer-School/pull/23

**Comment by teodutu** [comment]
It's not good yet. I have some more demos to add and to reword the explanations so as to start with the theory, not with the logger. Will do this by Friday. I'll also update the contents to work with the new layout.

**Comment by teodutu** [comment]
@RazvanN7 Now it should be alright. Please have another look.

**Comment by teodutu** [comment]
Not sure why the action is failing. The error looks unrelated: `fatal: could not read Username for 'https://github.com/': No such device or address`. @edi33416, do you have any idea what's happening?

**Comment by edi33416** [comment]
> Not sure why the action is failing. The error looks unrelated: `fatal: could not read Username for 'https://github.com/': No such device or address`. @edi33416, do you have any idea what's happening?

Yes. PR’s from forks cannot access secrets of the current repo. It’s a security feature designed to protect a repo from pr’s that might try to leak something or access a something.

You need to make a branch in the current repo and change the base of the pr to the newly created banch. That will run the action successfully

**Comment by teodutu** [comment]
That didn't work or I didn't understand what you meant. I think what needs to happen is get this PR merged into the branch and from there create another PR into `master`, right?

**Comment by edi33416** [comment]
> That didn't work or I didn't understand what you meant. I think what needs to happen is get this PR merged into the branch and from there create another PR into `master`, right?

The base branch should remain `master`.
You need to create the `rework-design-by-introspection` in this repo and then change the PR such that you want to merge the `rework-design-by-introspection` from this repo into `master`.

**Pull request closed without merging**